src/Features/CSharp/Portable/CodeRefactorings/ConvertToRecord/CSharpConvertToRecordRefactoringProvider.cs

```diff
@@ -0,0 +1,780 @@
+            if (bottom is IParameterReferenceOperation parameterReference)
+            {
+                return parameterReference.Parameter;
+            }
+            return null;
+        }
+
+        /// <summary>
+        /// Matches constructors where each statement simply assigns one of the provided parameters to one of the provided properties
+        /// with no duplicate assignment or any other type of statement
+        /// </summary>
+        /// <param name="constructor">Constructor Declaration</param>
+        /// <param name="properties">Properties expected to be assigned (would be replaced with positional constructor)</param>
+        /// <param name="parameters">Constructor parameters</param>
+        /// <returns>Whether the constructor body matches the pattern described</returns>
+        private static bool IsSimpleConstructor(
+            ConstructorDeclarationSyntax constructor,
+            ImmutableArray<PropertyDeclarationSyntax> properties,
+            SeparatedSyntaxList<ParameterSyntax> parameters,
+            SemanticModel semanticModel,
+            CancellationToken cancellationToken)
+        {
+            if (semanticModel.GetOperation(constructor, cancellationToken) is not IConstructorBodyOperation operation)
+            {
+                return false;
+            }
+
+            var body = operation.BlockBody ?? operation.ExpressionBody;
+
+            // We expect the constructor to have exactly one statement per parameter,
+            // where the statement is a simple assignment from the parameter to the property
+            if (body == null || body.Operations.Length != parameters.Count)
+            {
+                return false;
+            }
+
+            var propertyNames = properties.SelectAsArray(p => p.Identifier.ToString());
+            var propertyNamesAlreadyAssigned = new HashSet<string>();
+            var parameterNames = parameters.SelectAsArray(p => p.Identifier.ToString());
+
+            foreach (var bodyOperation in body.Operations)
+            {
+                if (bodyOperation is IExpressionStatementOperation
+                    {
+                        Operation: ISimpleAssignmentOperation
+                        {
+                            Target: IPropertyReferenceOperation propertyReference,
+                            Value: IParameterReferenceOperation parameterReference
+                        }
+                    })
+                {
+                    var propertyName = propertyReference.Property.Name;
+                    var propertyIndex = propertyNames.IndexOf(propertyName);
+                    if (propertyIndex != -1 &&
+                        !propertyNamesAlreadyAssigned.Contains(propertyName) &&
+                        // make sure the index of the property we assign as it would be placed in the primary constructor
+                        // matches the current index of the parameter we use for the explicit constructor
```

Comment:
So in the case of a single constructor that is doing the assignment of everything, wouldn't we just want to reorder the properties to match the parameter order?

Reply:
Yes, and I'll address this in a follow-up PR. However, our condition currently checks for sequence equality of the parameters to the properties, so currently this wouldn't even be checking constructors that have a different order of parameters.